docs/cudf/source/developer_guide/benchmarking.md

```diff
@@ -0,0 +1,288 @@
+We often want to generate quantitative comparisons, so we need to make that as easy as possible.
+Our benchmarks support this by setting the environment variable `CUDF_BENCHMARKS_USE_PANDAS`.
+When this variable is detected, all benchmarks will automatically be run using pandas instead of cuDF.
+Therefore, comparisons can easily be generated by simply running the benchmarks twice,
+once with the variable set and once without.
+
+```{note}
+`CUDF_BENCHMARKS_USE_PANDAS` effectively remaps `cudf` to `pandas` and `cupy` to `numpy`.
+It does so by aliasing these modules in `common.config.py`.
+This aliasing is why it is critical for developers to import these packages from `config.py`.
+```
+
+(testing)=
+
+## Testing
+
+Benchmarks need to be kept up to date with API changes in cuDF.
+However, we cannot simply run benchmarks in CI.
+Doing so would consume too many resources, and it would significantly slow down the development cycle
+
+To balance these issues, our benchmarks also support running in "testing" mode.
+To do so, developers can set the `CUDF_BENCHMARKS_DEBUG_ONLY` environment variable.
+When benchmarks are run with this variable, all data sizes are set to a minimum and the number of sizes are reduced.
+Our CI testing takes advantage of this to ensure that benchmarks remain valid code.
+
+```{note}
+The objects provided by `benchmark_with_object` respect the `NUM_ROWS` and `NUM_COLS` defined in `common/config.py`.
+`CUDF_BENCHMARKS_DEBUG_ONLY` works by conditionally redefining these values.
+This is why it is crucial for developers to use these variables when defining custom fixtures or cases.
+```
+
+## Profiling
+
+Although not strictly part of our benchmarking suite, profiling is a common need so we provide some guidelines here.
+There are two easy ways to profile benchmarks:
+1. The [`pytest-profiling`](https://github.com/man-group/pytest-plugins/tree/master/pytest-profiling) plugin.
+2. The [`py-spy`](https://github.com/benfred/py-spy) package.
+
+Using the former is as simple as adding the `--profile` (or `--profile-svg`) arguments to your `pytest` invocation.
+The latter requires instead invoking pytest from py-spy, like so:
+```
+py-spy record -- pytest bench_foo.py
+```
+Depending on exactly what information you need, your mileage may vary with each one.
+Developers should try both and see what works for their workflows.
+
+(advancedtopics)=
+
+## Advanced Topics
+
+This section discusses some underlying details of how cuDF benchmarks work.
+They are not usually necessary for typical developers or benchmark writers.
+This information is primarily for developers looking to extend the types of objects that can be easily benchmarked.
+
+### Understanding `benchmark_with_object`
+
+Under the hood, `benchmark_with_object` is made up of two critical pieces, fixture unions and some decorator magic.
+
+#### Fixture unions
+
+Fixture unions are a feature of `pytest_cases`.
```

Comment:
Link to pytest-cases docs?

Reply:
I didn't originally because it's already linked in the parametrization section, but since someone could definitely read this without reading that I agree that it's beneficial to add here too.